art/defences/preprocessor/spatial_smoothing_pytorch.py

```diff
@@ -0,0 +1,213 @@
+
+| Paper link: https://arxiv.org/abs/1704.01155
+
+
+| Please keep in mind the limitations of defences. For more information on the limitations of this defence,
+    see https://arxiv.org/abs/1803.09868 . For details on how to evaluate classifier security in general, see
+    https://arxiv.org/abs/1902.06705
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+from kornia.filters.kernels import get_binary_kernel2d
+import torch.nn.functional as F
+import torch.nn as nn
+
+import logging
+from typing import Optional, Tuple
+
+import numpy as np
+import torch
+from kornia.filters import MedianBlur
+
+from art.defences.preprocessor.spatial_smoothing import SpatialSmoothing
+from art.utils import Deprecated, deprecated_keyword_arg
+
+logger = logging.getLogger(__name__)
+
+
+class MedianBlurCustom(MedianBlur):
+    """
+    An ongoing effort to reproduce the median blur function in SciPy.
+    """
+
+    def __init__(self, kernel_size: Tuple[int, int]) -> None:
+        super().__init__(kernel_size)
+
+        # Half-pad the input so that the output keeps the same shape.
+        # * center pixel located lower right
+        half_pad = [k % 2 == 0 for k in kernel_size]
+        self.p2d = (self.padding[-1] + half_pad[-1], self.padding[-1],
+                    self.padding[-2] + half_pad[-2], self.padding[-2])
+        # PyTorch requires Padding size should be less than the corresponding input dimension,
+
+    def forward(self, input: torch.Tensor):  # type: ignore
+        if not torch.is_tensor(input):
+            raise TypeError("Input type is not a torch.Tensor. Got {}".format(type(input)))
+        if not len(input.shape) == 4:
+            raise ValueError("Invalid input shape, we expect BxCxHxW. Got: {}".format(input.shape))
+        # prepare kernel
+        b, c, h, w = input.shape
+        kernel: torch.Tensor = self.kernel.to(input.device).to(input.dtype)
+        # map the local window to single vector
+
+        _input = input.reshape(b * c, 1, h, w)
+        if input.dtype == torch.int64:
+            # "reflection_pad2d" not implemented for 'Long'
+            # "reflect" in scipy.ndimage.median_filter has no equivalance in F.pad.
+            # "reflect" in PyTorch maps to "mirror" in scipy.ndimage.median_filter.
+            _input = _input.to(torch.float32)
+            _input = F.pad(_input, self.p2d, "reflect")
+            _input = _input.to(torch.int64)
+        else:
+            _input = F.pad(_input, self.p2d, "reflect")
+
+        features: torch.Tensor = F.conv2d(_input, kernel, stride=1)
+        features = features.view(b, c, -1, h, w)  # BxCx(K_h * K_w)xHxW
+
+        # compute the median along the feature axis
+        # * torch.median(), if window size even, use smaller value (e.g. median(4,5)=4)
+        median: torch.Tensor = torch.median(features, dim=2)[0]
+        return median
+
+
+class SpatialSmoothingPyTorch(SpatialSmoothing):
```

Comment:
I'm wondering if we should introduce a new base class `PreprocessorPyTorch`. That would allow us to check if the defence is PyTorch specific by checking its type instead of checking for the method `forward` and it would also allow us to introduce framework-specific preprocessing defences for other frameworks later on.

Reply:
Sure.